File: mobile-app/lib/models/attendance.dart
enum AttendanceStatus {
  present,
  absent,
  late,
  excused,
}

extension AttendanceStatusExtension on AttendanceStatus {
  String get displayName {
    switch (this) {
      case AttendanceStatus.present:
        return 'Present';
      case AttendanceStatus.absent:
        return 'Absent';
      case AttendanceStatus.late:
        return 'Late';
      case AttendanceStatus.excused:
        return 'Excused';
    }
  }

  String get value {
    switch (this) {
      case AttendanceStatus.present:
        return 'present';
      case AttendanceStatus.absent:
        return 'absent';
      case AttendanceStatus.late:
        return 'late';
      case AttendanceStatus.excused:
        return 'excused';
    }
  }

  static AttendanceStatus fromString(String value) {
    switch (value.toLowerCase()) {
      case 'present':
        return AttendanceStatus.present;
      case 'absent':
        return AttendanceStatus.absent;
      case 'late':
        return AttendanceStatus.late;
      case 'excused':
        return AttendanceStatus.excused;
      default:
        return AttendanceStatus.absent;
    }
  }
}

class Attendance {
  final int id;
  final int studentId;
  final int sessionId;
  final AttendanceStatus status;
  final DateTime timestamp;
  final double? latitude;
  final double? longitude;
  final String? deviceId;
  final String? notes;
  final bool isValid;
  final String? validationErrors;
  final DateTime createdAt;
  final DateTime updatedAt;

  Attendance({
    required this.id,
    required this.studentId,
    required this.sessionId,
    required this.status,
    required this.timestamp,
    this.latitude,
    this.longitude,
    this.deviceId,
    this.notes,
    required this.isValid,
    this.validationErrors,
    required this.createdAt,
    required this.updatedAt,
  });

  bool get isLate => status == AttendanceStatus.late;
  bool get isPresent => status == AttendanceStatus.present;
  bool get isAbsent => status == AttendanceStatus.absent;
  bool get isExcused => status == AttendanceStatus.excused;

  factory Attendance.fromJson(Map<String, dynamic> json) {
    return Attendance(
      id: json['id'],
      studentId: json['student_id'],
      sessionId: json['session_id'],
      status: AttendanceStatusExtension.fromString(json['status']),
      timestamp: DateTime.parse(json['timestamp']),
      latitude: json['latitude']?.toDouble(),
      longitude: json['longitude']?.toDouble(),
      deviceId: json['device_id'],
      notes: json['notes'],
      isValid: json['is_valid'] ?? true,
      validationErrors: json['validation_errors'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'student_id': studentId,
      'session_id': sessionId,
      'status': status.value,
      'timestamp': timestamp.toIso8601String(),
      'latitude': latitude,
      'longitude': longitude,
      'device_id': deviceId,
      'notes': notes,
      'is_valid': isValid,
      'validation_errors': validationErrors,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}

class AttendanceSummary {
  final int totalSessions;
  final int presentCount;
  final int absentCount;
  final int lateCount;
  final int excusedCount;
  final double attendanceRate;
  final DateTime? lastAttendanceDate;

  AttendanceSummary({
    required this.totalSessions,
    required this.presentCount,
    required this.absentCount,
    required this.lateCount,
    required this.excusedCount,
    required this.attendanceRate,
    this.lastAttendanceDate,
  });

  factory AttendanceSummary.fromJson(Map<String, dynamic> json) {
    return AttendanceSummary(
      totalSessions: json['total_sessions'] ?? 0,
      presentCount: json['present_count'] ?? 0,
      absentCount: json['absent_count'] ?? 0,
      lateCount: json['late_count'] ?? 0,
      excusedCount: json['excused_count'] ?? 0,
      attendanceRate: (json['attendance_rate'] ?? 0).toDouble(),
      lastAttendanceDate: json['last_attendance_date'] != null
          ? DateTime.parse(json['last_attendance_date'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'total_sessions': totalSessions,
      'present_count': presentCount,
      'absent_count': absentCount,
      'late_count': lateCount,
      'excused_count': excusedCount,
      'attendance_rate': attendanceRate,
      'last_attendance_date': lastAttendanceDate?.toIso8601String(),
    };
  }
}

class AttendanceRecord {
  final Attendance attendance;
  final String courseName;
  final String courseCode;
  final String sessionTitle;
  final DateTime sessionDate;
  final String lecturerName;

  AttendanceRecord({
    required this.attendance,
    required this.courseName,
    required this.courseCode,
    required this.sessionTitle,
    required this.sessionDate,
    required this.lecturerName,
  });

  factory AttendanceRecord.fromJson(Map<String, dynamic> json) {
    return AttendanceRecord(
      attendance: Attendance.fromJson(json['attendance']),
      courseName: json['course_name'] ?? '',
      courseCode: json['course_code'] ?? '',
      sessionTitle: json['session_title'] ?? '',
      sessionDate: DateTime.parse(json['session_date']),
      lecturerName: json['lecturer_name'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'attendance': attendance.toJson(),
      'course_name': courseName,
      'course_code': courseCode,
      'session_title': sessionTitle,
      'session_date': sessionDate.toIso8601String(),
      'lecturer_name': lecturerName,
    };
  }
}

File: mobile/app/services/models/attendance.ts
/**
 * Attendance record model
 */
export interface AttendanceRecord {
    id: number;
    student: number; // User ID
    session: number; // CourseSession ID
    course: number; // Course ID
    timestamp: string;
    latitude?: number;
    longitude?: number;
    device_fingerprint?: string;
    qr_code_data?: string;
    status: 'present' | 'late' | 'absent';
    is_verified: boolean;
    marked_by?: number; // User ID who marked the attendance
    notes?: string;
}

/**
 * Attendance summary for a student
 */
export interface AttendanceSummary {
    student_id: number;
    student_name: string;
    course_id: number;
    course_name: string;
    total_sessions: number;
    present_count: number;
    late_count: number;
    absent_count: number;
    attendance_percentage: number;
}

/**
 * Attendance marking request
 */
export interface AttendanceMarkRequest {
    qr_code: string;
    latitude?: number;
    longitude?: number;
    device_fingerprint?: string;
}

/**
 * QR code verification response
 */
export interface QRVerificationResponse {
    valid: boolean;
    session_id?: number;
    course_id?: number;
    instructor_name?: string;
    session_title?: string;
    expires_at?: string;
    error?: string;
}

/**
 * Attendance statistics
 */
export interface AttendanceStats {
    total_students: number;
    present_today: number;
    attendance_rate: number;
    course_id?: number;
    session_id?: number;
}

/**
 * Check if attendance record is present
 */
export function isPresent(record: AttendanceRecord): boolean {
    return record.status === 'present';
}

/**
 * Check if attendance record is late
 */
export function isLate(record: AttendanceRecord): boolean {
    return record.status === 'late';
}

/**
 * Check if attendance record is absent
 */
export function isAbsent(record: AttendanceRecord): boolean {
    return record.status === 'absent';
}

/**
 * Calculate attendance percentage
 */
export function calculateAttendancePercentage(
    present: number,
    late: number,
    total: number
): number {
    if (total === 0) return 0;
    // Late counts as half attendance
    const effectiveAttendance = present + (late * 0.5);
    return Math.round((effectiveAttendance / total) * 100);
}

/**
 * Get attendance status display name
 */
export function getAttendanceStatusDisplay(status: 'present' | 'late' | 'absent'): string {
    switch (status) {
        case 'present':
            return 'Present';
        case 'late':
            return 'Late';
        case 'absent':
            return 'Absent';
        default:
            return 'Unknown';
    }
}

/**
 * Get attendance status color (for UI)
 */
export function getAttendanceStatusColor(status: 'present' | 'late' | 'absent'): string {
    switch (status) {
        case 'present':
            return '#4CAF50'; // Green
        case 'late':
            return '#FF9800'; // Orange
        case 'absent':
            return '#F44336'; // Red
        default:
            return '#9E9E9E'; // Grey
    }
}

/**
 * Check if QR code is expired
 */
export function isQRCodeExpired(expiresAt?: string): boolean {
    if (!expiresAt) return false;
    const expiryDate = new Date(expiresAt);
    const now = new Date();
    return now > expiryDate;
}
